store/cachemulti/store.go

```diff
@@ -25,77 +25,59 @@ const storeNameCtxKey = "store_name"
 	traceContext types.TraceContext
 
 	listeners map[types.StoreKey][]types.WriteListener
 }
 
-var _ types.CacheMultiStore = Store{}
+var _ types.CacheMultiStore = (*Store)(nil)
 
 // NewFromKVStore creates a new Store object from a mapping of store keys to
 // CacheWrapper objects and a KVStore as the database. Each CacheWrapper store
 // is a branched store.
 func NewFromKVStore(
 	store types.KVStore, stores map[types.StoreKey]types.CacheWrapper,
 	keys map[string]types.StoreKey, traceWriter io.Writer, traceContext types.TraceContext,
-	listeners map[types.StoreKey][]types.WriteListener,
-) Store {
-	if listeners == nil {
-		listeners = make(map[types.StoreKey][]types.WriteListener)
-	}
-	cms := Store{
-		db:           cachekv.NewStore(store),
-		stores:       make(map[types.StoreKey]types.CacheWrap, len(stores)),
-		keys:         keys,
-		traceWriter:  traceWriter,
-		traceContext: traceContext,
-		listeners:    listeners,
-	}
-
-	for key, store := range stores {
-		if cms.TracingEnabled() {
-			tctx := cms.traceContext.Clone().Merge(types.TraceContext{
-				storeNameCtxKey: key.Name(),
-			})
-
-			store = tracekv.NewStore(store.(types.KVStore), cms.traceWriter, tctx)
-		}
-		if cms.ListeningEnabled(key) {
-			store = listenkv.NewStore(store.(types.KVStore), key, listeners[key])
-		}
-		cms.stores[key] = cachekv.NewStore(store.(types.KVStore))
+) *Store {
+	cms := &Store{
+		db:             cachekv.NewStore(store),
+		originalStores: stores,
+		keys:           keys,
+		traceWriter:    traceWriter,
+		traceContext:   traceContext,
 	}
 
+	cms.stores = cms.recreateStores()
 	return cms
 }
 
 // NewStore creates a new Store object from a mapping of store keys to
 // CacheWrapper objects. Each CacheWrapper store is a branched store.
 func NewStore(
 	db dbm.DB, stores map[types.StoreKey]types.CacheWrapper, keys map[string]types.StoreKey,
-	traceWriter io.Writer, traceContext types.TraceContext, listeners map[types.StoreKey][]types.WriteListener,
-) Store {
-	return NewFromKVStore(dbadapter.Store{DB: db}, stores, keys, traceWriter, traceContext, listeners)
+	traceWriter io.Writer, traceContext types.TraceContext,
+) *Store {
+	return NewFromKVStore(dbadapter.Store{DB: db}, stores, keys, traceWriter, traceContext)
 }
 
-func newCacheMultiStoreFromCMS(cms Store) Store {
+func newCacheMultiStoreFromCMS(cms *Store) *Store {
 	stores := make(map[types.StoreKey]types.CacheWrapper)
 	for k, v := range cms.stores {
 		stores[k] = v
 	}
 
-	// don't pass listeners to nested cache store.
-	return NewFromKVStore(cms.db, stores, nil, cms.traceWriter, cms.traceContext, nil)
+	return NewFromKVStore(cms.db, stores, nil, cms.traceWriter, cms.traceContext)
 }
 
 // SetTracer sets the tracer for the MultiStore that the underlying
 // stores will utilize to trace operations. A MultiStore is returned.
-func (cms Store) SetTracer(w io.Writer) types.MultiStore {
+func (cms *Store) SetTracer(w io.Writer) types.MultiStore {
 	cms.traceWriter = w
 	return cms
 }
```

Comment:
By changing the method receiver from a value to a pointer, you change the fundamental behavior of the code. Previously, this method had no effect on the `traceWriter` field of the Store; now, it overwrites that field. That's certainly unexpected by callers, and will likely break consumer code in subtle and surprising ways. What motivates the change?

Reply:
Maintainers, please review this PR very carefully, the risk is much larger than it may appear to be.

Reply:
> By changing the method receiver from a value to a pointer, you change the fundamental behavior of the code. Previously, this method had no effect on the `traceWriter` field of the Store; now, it overwrites that field. That's certainly unexpected by callers, and will likely break consumer code in subtle and surprising ways. The same thing is true of basically every other similar change in the PR. What motivates those changes?

make sense, does that mean currently [this code](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/options.go#L196) has no effect?

Reply:
make sense, I'll find a way to not modify the method receiver.

Reply:
# What motivates those changes?

The motivation is explained in the issue #13457, specifically, [this comment](https://github.com/cosmos/cosmos-sdk/issues/13457#issuecomment-1271396206).
Basically, the current state listener is pretty much broken.

Reply:
I also updated the adr-038, ready for review now.